bundles/org.openhab.binding.argoclima/src/main/java/org/openhab/binding/argoclima/internal/configuration/ArgoClimaConfigurationRemote.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * @author Mateusz Bronk - Initial contribution
+ */
+@NonNullByDefault
+public class ArgoClimaConfigurationRemote extends ArgoClimaConfigurationBase {
+
+    /**
+     * The duration after which the device would be considered non-responsive (and taken OFFLINE)
+     */
+    public static final Duration LAST_SEEN_UNAVAILABILITY_THRESHOLD = Duration.ofMinutes(20);
+
+    /**
+     * Argo configuration parameters specific to remote connection
+     * These names are defined in thing-types.xml and get injected on instantiation
+     * through {@link org.openhab.core.thing.binding.BaseThingHandler#getConfigAs getConfigAs}
+     */
+    private String username = "";
+    private String password = "";
+
+    /**
+     * Get the username (login) to use in authenticating to Argo server
+     *
+     * @return username (as configured by the user)
+     */
+    public String getUsername() {
+        return this.username;
+    }
+
+    /**
+     * Get the masked password used in authenticating to Argo server (for logging)
+     *
+     * @implNote Password length is preserved (which may be considered a security weakness, but is useful for
+     *           troubleshooting and given state of Argo API's security... likely is an overkill already :)
+     * @return {@code ***}-masked string instead of the same length as configured password
+     */
+    private final String getPasswordMasked() {
+        return this.password.replaceAll(".", "*");
+    }
+
+    /**
+     * Get MD5 hash of the configured password (for Basic auth)
+     *
+     * @return MD5 hash of password
+     * @throws ArgoConfigurationException In case MD5 is not available in the security provider
+     */
+    public String getPasswordMD5Hash() throws ArgoConfigurationException {
+        MessageDigest md;
+        try {
+            md = MessageDigest.getInstance("MD5");
+            md.update(password.getBytes());
+            byte[] digest = md.digest();
+            return DatatypeConverter.printHexBinary(digest).toLowerCase();
+        } catch (NoSuchAlgorithmException e) {
+            throw ArgoConfigurationException.forInvalidParamValue(ArgoClimaBindingConstants.PARAMETER_PASSWORD,
+                    getPasswordMasked(), i18nProvider, e); // User-provided value is likely NOT at fault, but using this
+                                                           // exception for generic error messaging (cause will be
+                                                           // displayed anyway)
+        }
+    }
+
+    @Override
+    protected String getExtraFieldDescription() {
+        return String.format("username=%s, password=%s", username, getPasswordMasked());
+    }
+
+    @Override
+    protected void validateInternal() throws ArgoConfigurationException {
+        if (username.isBlank()) {
+            throw ArgoConfigurationException.forEmptyRequiredParam(ArgoClimaBindingConstants.PARAMETER_USERNAME,
+                    i18nProvider);
```

Comment:
You can get rid if the dependency of the i18nprovider in the configuration classes. The configuraiton class should not be responsible for translation. only for holding the configuration and logic for validation of it.

If you throw ArgoConfigurationException("specific i18n key for translation") the translation can be performed at the last possible place (when setting the thingstatus later on), see other comments.
that would also make the exception throwing simpler.

Reply:
TBH configuration is not responsible for translation, just passing through the provider, to cut on the boilerplate _(I initially was throwing exactly as you suggest, but it creates massive code duplication if you don't have self-contained/static exceptions - Notice these are not simple static strings, but include inner causes, possibly N-level deep and I want to keep that, otherwise I'm either degrading UX or writing an `if` ladder and redoing messages which runtime already has localized)._
It surely is NOT a ideal/golden way of doing it (cuts corners in a few cases when using localized cause message at throw site), but note these are only a few. Im most cases it **already** behaves as I think you describe (exception is full en-us and has a key in it... and only gets translated on request (at thing handler).

Aka. I see no harm in configuration class having the i18n provider to inject (poor man's dep injection)... rather the select **few** places calling `getLocalizedMessage` prematurely (which can be improved, though adding complexity)

Reply:
See other comment.